urllib3/sessionmanager.py

```diff
@@ -0,0 +1,78 @@
+from .util.retry import Retry, handle_redirect
+from .request import RequestMethods, Request
+from .poolmanager import PoolManager, ProxyManager
+from .util.context import Context
+
+
+class SessionManager(RequestMethods):
+    """
+    Allows arbitrary requests while maintaining browser-session-like
+    context across those requests.
+
+    :param headers:
+        Headers to include with all requests, unless other headers are
+        given explcitly
+
+    :param cookie_jar:
+        A predefined :class:`CookieJar` object to use in the session;
+        if not provided, a new one will be created.
+
+    :param \**pm_kwargs:
+        Additional paramaters to be passed to the encapsulated
+        :class:`.PoolManager` or :class:`.ProxyManager` instance.
+    """
+    def __init__(self, headers=None, cookie_jar=None, **pm_kwargs):
+
+        super(SessionManager, self).__init__(headers=headers)
+
+        # Detect if we should create a PoolManager or a ProxyManager
+        if pm_kwargs.get('proxy_url', None) is None:
```

Comment:
> That said, I would prefer to avoid this flow altogether and instead pass in a pre-made Manager to the constructor. 

Do you mind walking me through the thinking here? Writing this, I sort of was seeing `SessionManager` as abstracting the `PoolManager` and `ProxyManager` in the same way those classes abstract the `ConnectionPool`, so I want to make sure that my thinking lines up with yours when revisiting it.